packages/api/src/modules/providers/scdl/scdl.grant.parser.ts

```diff
@@ -46,27 +93,79 @@ export default class ScdlGrantParser {
 
         const headerRow = page[rowOffset] as string[];
         console.log("Map rows to entities...");
-        const data = page.slice(rowOffset + 1).map(row => ParserHelper.linkHeaderToData(headerRow, row));
-        return ScdlGrantParser._filterValidEntities(data);
+        const data = page.slice(rowOffset + 1).map(row => GenericParser.linkHeaderToData(headerRow, row));
+        return ScdlGrantParser.convertValidateData(data);
     }
 
-    static _filterValidEntities(parsedChunk) {
+    protected static convertValidateData(parsedChunk): {
+        entities: ScdlStorableGrant[];
+        errors: ParsedDataWithProblem[];
+    } {
         const storableChunk: ScdlStorableGrant[] = [];
         const invalidEntities: Partial<ScdlStorableGrant>[] = [];
+        const errors: ParsedDataWithProblem[] = [];
 
         for (const parsedData of parsedChunk) {
-            const entity = ParserHelper.indexDataByPathObject(SCDL_MAPPER, parsedData) as unknown as ScdlParsedGrant;
-            if (this.isGrantValid(entity)) storableChunk.push({ ...entity, __data__: parsedData });
-            else {
-                invalidEntities.push(entity);
-            }
+            const {
+                entity,
+                annotations,
+                errors: errorsEntity,
+            } = ScdlGrantParser.indexDataByPathAndAnnotate<string, ScdlStorableGrant>(SCDL_MAPPER, parsedData);
+            errors.push(...errorsEntity);
+
+            // validates and saves annotated errors
+            const validation = this.isGrantValid(entity as ScdlStorableGrant, annotations);
+            if (validation.valid)
+                storableChunk.push({ ...this.cleanOptionalFields(entity as ScdlStorableGrant), __data__: parsedData });
+            else invalidEntities.push(entity);
+            validation?.problems?.map((pb: Problem) => errors.push({ ...parsedData, ...pb }));
         }
 
         if (invalidEntities.length) {
             console.log(`WARNING : ${invalidEntities.length} entities invalid`);
             console.log(`Here some of them: `, invalidEntities.slice(0, 3));
         }
 
-        return storableChunk;
+        return { entities: storableChunk, errors };
+    }
+
+    /*
+    equivalent of GenericParser.indexDataByPathObject but keeping in memory the initial label
+    could be extracted to GenericParser but at this point (#2387) it seems overkill
+     */
+    static indexDataByPathAndAnnotate<Tin extends BeforeAdaptation, Tout = DefaultObject>(
```

Comment:
Qu'entends tu par "Annotate" ?

Reply:
Les annotations c'est l'objet `{[key]: {value, keyPath}}`

Reply:
plutôt `{ej: { value: "09245253", keyPath: ["path", "to", "ej"]}` mais c'est l'idée